addons/website_sale/controllers/main.py

```diff
@@ -636,6 +638,18 @@ def checkout_redirection(self, order):
         if tx and tx.state != 'draft':
             return request.redirect('/shop/payment/confirmation/%s' % order.id)
 
+        # Force login if tokenization is required
+        if request.env['payment.acquirer'].sudo()._is_tokenization_required(
+            sale_order_id=order.id
+        ):
+            user_sudo = request.env.user
+            # The user needs to be logged if tokenization is required.
+            if user_sudo._is_public():
+                return request.redirect(
+                    # Escape special characters to avoid loosing original params when redirected
+                    f'/web/login?redirect={urllib.parse.quote(request.httprequest.full_path)}'
+                )
```

Comment:
Are you sure this is needed for e-commerce flows ?

Reply:
Yes because we don't save tokens if the user is not logged in.